File: binary-search/BinarySearch.java
public class BinarySearch {
  public static int binarySearch(int[] array, int target) {
    return binarySearch(array, target, 0, array.length - 1);
  }
  public static int binarySearch(int[] array, int target, int left, int right) {
    if(left > right) { return - 1; }
    int middle  = (left + right) / 2;
    int potentialMatch = array[middle];
    if(target == potentialMatch) {
      return middle;
    } else if (target < potentialMatch) {
      return binarySearch(array, target, left, middle - 1);
    } else {
      return binarySearch(array, target, middle + 1, right);
    }
  }

  public static int binarySearchTwo(int[] array, int target) {
    return binarySearchTwo(array, target, 0, array.length - 1);
  }

  public static int binarySearchTwo(int[] array, int target, int start, int end) {
    while(start <= end) {
      int middle = (start + end) / 2;
      int potentialMatch = array[middle];
      if(target == potentialMatch) {
        return middle;
      } else if (target < potentialMatch) {
          end = middle - 1;
      } else {
          start = middle + 1;
      }
    }
    return -1;
  }
}

File: binary-search/BinarySearch.swift
import Foundation

class BinarySearch {
    func binarySearchOne(_ array: [Int], _ target: Int) -> Int {
        var left = 0
        var right = array.count - 1;
        return binarySearchIterative(array, target, &left, &right)
    }
    
    func binarySearchIterative(_ array: [Int], _ target: Int, _ left: inout Int, _ right: inout Int) -> Int {
        while left <= right {
            var middle = (left + right) / 2;
            var potentialMatch = array[middle]
            if(target == potentialMatch) {
                return middle
            } else if (target < potentialMatch) {
                right = middle - 1
            } else {
                left = middle + 1
            }
        }
        return -1
    }

    func binarySearchTwo(_ array: [Int], _ target: Int) -> Int {
        var left = 0
        var right = array.count - 1;
        return binarySearchRecursive(array, target, &left, &right)
    }

    func binarySearchRecursive(_ array: [Int], _ target: Int, _ left: inout Int, _ right: inout Int) -> Int {
        if left > right { return -1 }
        let middle = (left + right) / 2;
        let potentialMatch = array[middle]
        if(target == potentialMatch) {
            return middle
        } else if target < potentialMatch {
            right = middle - 1
            return binarySearchRecursive(array, target, &left, &right)
        } else {
            left = middle + 1
            return binarySearchRecursive(array, target, &left, &right)
        }
    }
}